consensus/core/src/block_verifier.rs

```diff
@@ -185,20 +188,44 @@ impl BlockVerifier for SignedBlockVerifier {
     fn check_ancestors(
         &self,
         block: &VerifiedBlock,
-        ancestors: &[VerifiedBlock],
+        ancestors: &[Option<VerifiedBlock>],
+        gc_enabled: bool,
+        gc_round: Round,
     ) -> ConsensusResult<()> {
-        assert_eq!(block.ancestors().len(), ancestors.len());
-        // This checks the invariant that block timestamp >= max ancestor timestamp.
-        let mut max_timestamp_ms = BlockTimestampMs::MIN;
-        for (ancestor_ref, ancestor_block) in block.ancestors().iter().zip(ancestors.iter()) {
-            assert_eq!(ancestor_ref, &ancestor_block.reference());
```

Comment:
I think you just need to add this line before the assert, something like:
```
let Some(ancestor_block) = ancestor_block else {
   assert!(gc_enabled && ancestor_ref.round <= gc_round);
   continue;
};
```

Reply:
Yep can do, but I feel safer at this point just separating those two paths completely. The timestamp logic will get refactored anyways.